### PR TITLE
networkd: allow to retain configs even if carrier is lost

### LINKED ISSUE #9111: DHCP client fails to negotiate with Cox Cable
All versions for past 3-4 years.

**Used distribution**
 > Arch Linux

**Expected behaviour you didn't see**
 > It should get a valid IPv4 Address without manual intervention.

**Unexpected behaviour you saw**
 > Negotiation fails.... it gets a temporary address from the modem, and Cox tries to go through a couple redirections to give the proper IP... giving first a 10. address and then assigning a 172. Address afterward.

**Steps to reproduce the problem**
Unfortunatly, this probably requires also having the same network provider and perhaps even the same city as me to reproduce...

Attached networkd debug log, and a tcpdump (mycap6) at the same time.  redid a negitation on windows and attached that wireshark capture as pcap.(windows.zip)

[mycap6.zip](https://github.com/systemd/systemd/files/2046247/mycap6.zip)
[networkd.debug.log](https://github.com/systemd/systemd/files/2046248/networkd.debug.log)
[windows.zip](https://github.com/systemd/systemd/files/2046266/windows.zip)

What I have to do to get a valid IP is clone the MAC address from the linux box's external adapter to a windows system, and then use windows to negotiate the IP, then I can re-plug in the cable to the linux box, and the IP is able to be received by the linux box.

This has been the case ever since I installed this Arch system which uses systemd; which was several years ago.

]

### PR DESCRIPTION
**Pull request by ssahani**
When there is bad link in the network the carrier goes up/down.
This makes networkd stops all the clients and drop config.
But if the remote router/dhcpserver running a prevention
of DHCP Starvation attack or DHCP Flood attack it does not allow
networkd to take a DHCP lease resulting failure in configuration.
This patch allows to keep the client running and keep the conf
also for this scenario.

Closes #9111

### CODE REVIEW

File: man/systemd.network.xml [hunk 1]
```diff
@@ -779,6 +779,15 @@
           </para>
         </listitem>
       </varlistentry>
+      <varlistentry>
+        <term><varname>IgnoreCarrierLoss=</varname></term>
+        <listitem>
+          <para>A boolean. Allows networkd to retain the configuration even if no carrier.
+          Defaults to false.
```
Comment (keszybz) [inline man/systemd.network.xml]: This seems to be OK, but the documentation needs to be better. "Retain the configuration" makes it sound like networkd drops user configuration when there is carrier loss, which of course it doesn't. I guess you mean "retain DHCP client state" or something like that. Please adjust this text to describe what is being retained exactly.

Comment (ssahani) [reply man/systemd.network.xml]: Confs = DHCP and static both are retained .

**Comment by yuwata** [comment]
@ssahani Could you rebase this and update man page?

**Comment by yuwata** [comment]
Also, adding a testcase is welcome.

File: man/systemd.network.xml (hunk 1 above)
Comment (yuwata) [reply man/systemd.network.xml]: So, it is not necessary to list all entries retained by the setting, but I think listing some of them is helpful. Both static and dynamic addresses, routes, routing rules, neighbors,...

**Comment by ssahani** [comment]
@yuwata updated thanks ! not sure how we manipulate the carrier in test case.

**Comment by yuwata** [comment]
@ssahani Thank you. I've slightly modify your patch, and add testcases. PTAL.

**Comment by ssahani** [comment]
LGTM thanks !

**Comment by yuwata** [comment]
fuzz-network/directives.network is updated. I hope CIs will pass now.

**Comment by amishmm** [comment]
I have not tested but what happens in following case?

1) networkd starts and link is up and link_carrier_gained() is called - which calls link_acquire_conf() and link_enter_set_addresses()
2) Now link goes down --- nothing happens becuase IgnoreCarrierLoss is set.
3) Now link is up again --- does it try to acquire conf and set addresses again?

So if link goes down 10 times .. will it try to acquire conf and set address as many times? If yes, then its not necessary. Please see patch in #7403 

Also I am sorry but I fail to understand when there was already a PR by me and @ssahani knew about it too and he had blocked my PR too by denying it for reasons he mentioned in #7403 

Now he has done exactly same thing in his PR. This definitely discourages people to send PR because I have spent weeks developing and pushing for this feature.

But anyway.

**Comment by yuwata** [comment]
Sorry for merging this without commenting in #7403 or to you.

> So if link goes down 10 time and up again .. will it try to acquire conf and set address as many times? If yes, then its not necessary.

It may not be necessary, though I am not sure there is no corner case, but I think reconfiguring interface is harmless. Of course, we can optimize the flow, but the operation seems not so heavy, no?

----
> Also I am sorry but I fail to understand when there was already a PR by me and @ssahani knew about it too and he had blocked my PR too by denying it for reasons he mentioned in #7403

> Now he has done exactly same thing in his PR. This definitely discourages people to send PR because I have spent weeks developing and pushing for this feature.

At least in the case of these two PRs, I added test cases for this PR, and seems to work fine. Of course I believe you tested your PR well by yourself, but at that time, no visible tests for #7403 was included...

----
Anyway, follow-up commits for this is welcome. Thank you!

**Comment by amishmm** [comment]
Yes I and many others had tested #7403 and reported back that it was working fine for them. Only difference this PR and my PR has is that my PR does it only for static IP cases and not DHCP.

But hey wait, I added that check for static IP condition because @ssahani himself asked me to do it only for static case and not DHCP.

Btw not having a test case can not be reason for denial of PR. And I was not told anywhere that my PR is on hold because its missing test case. (To be honest, I have no idea on how to create test case though)

**Anyway lets move on.**

**Coming back to this PR**... it probably has a race condition. I am saying probably because I am not completely aware on how systemd-networkd actually works. (i.e. it may not be race condition but i still have a doubt that there is)

What would happen if link is goes up-down-up rapidly **before link is configured**?

1. Link up --- configuration is going on but not yet complete (or partially configured)
2. Link down --- nothing happens because ignorecarrierloss is set.
3. Link up again --- what happens now? I am not sure. Will it restart from where it was partially configured or re-configure everything from scratch?

This may create a race condition or may leave link in an undefined state. (again not sure)

So as I have mentioned in my previous comment we need to check for LINK_STATE_CONFIGURED first and then "return 0" if IgnoreCarrierLoss is set. See patch in #7403 

i.e.

1. Link up --- configuration is going on but not yet complete (or partially configured)
2. Link down ---
- a) link not yet configured - go ahead unconfigure it (reverse whatever it configured)
- b) link configured - do nothing if ignorecarrierloss is set
3. Link up again --- proceed normally

I believe there is no harm done if we check LINK_STATE_CONFIGURED first before acting on IgnoreCarrierLoss.

**Comment by yuwata** [comment]
> testcases

Now we have systemd-networkd-tests.py and it is run by CentOS CI. Yeah it did not exist when you open the PR, unfortunately...

> What would happen if link is goes up-down-up rapidly **before link is configured**?
> 
>     1. Link up --- configuration is going on but not yet complete (or partially configured)
> 
>     2. Link down --- nothing happens because ignorecarrierloss is set.
> 
>     3. Link up again --- what happens now? I am not sure. Will it restart from where it was partially configured or re-configure everything from scratch?

Interesting. It is worth to add a test case for that.

**Comment by amishmm** [comment]
Please do it. I have no idea on how to create a test case.
> It may not be necessary, though I am not sure there is no corner case, but I think reconfiguring interface is harmless.

Please also add a test case on what happens if I have:
[RoutingPolicyRule]
To=8.8.8.8
Table=100

I believe with current PR - this is going to create duplicate rules everytime link goes down and back up.

Because:
1) Link up - acquire conf. rule added
2) Link down - untouched
3) Link up - acquire conf. rule added again?

Sorry if I am wrong in interpretation. As I am not sure if acquire conf will drop any existing conf or not.

**Comment by yuwata** [comment]
OK. Thank you for the suggestions.

**Comment by yuwata** [comment]
I've added a testcase you suggested in #11767, and it seems to work fine :-)

**Comment by amishmm** [comment]
Thank you. That was very quick. I have a query posted here: https://github.com/systemd/systemd/pull/11767/commits/ed061902de948c2f81f153cf2334fa5ed00cd65a#r258327549

**Pull request merged**